### PR TITLE
Add `bench test-progress` command for per-test partial-credit scoring

### PR DESCRIPTION
**Pull request by madmax983**
## Summary

Implements a new `bench test-progress` subcommand that computes partial-credit scores for every instance in a completed sweep, enabling gradient-based evaluation beyond the binary resolved/unresolved verdict. The command reads existing evaluator output and dataset metadata to produce per-instance metrics, sweep-wide aggregations, and identifies the most problematic tests across the sweep.

## Key Changes

- **New module `src/run/test_progress.rs`** (841 lines):
  - Core scoring logic: `compute_partial_credit_score()` formula that clamps `(pass_ratio - regress_ratio)` to `[-1.0, 1.0]`
  - Per-instance metrics computation: `compute_instance_metrics()` tracks FAIL_TO_PASS and PASS_TO_PASS test outcomes
  - Verdict bucket classification: Resolved, PartialProgress, NoProgress, Regressed, EvaluatorUnavailable
  - Report generation: Aggregates per-instance data into sweep-wide totals, identifies hot failing/regressed tests
  - Integration with `bench compare`: `test_progress_compare_section()` produces delta summaries between baseline and candidate sweeps
  - Text rendering: `render_text()` formats results as human-readable tables

- **CLI integration** (`src/cli/args.rs`, `src/cli/mod.rs`):
  - New `TestProgressCmd` struct with options: `--sweep`, `--format`, `--bucket`, `--hot-tests-n`, `--filter`, `--min-tests`
  - Wired into main command dispatcher to invoke `bench_test_progress()`
  - Integrated into `bench compare` output to show test-progress deltas

- **Comprehensive test suite** (`tests/bench_test_progress.rs`, 764 lines):
  - Unit tests for scoring formula and verdict bucket logic
  - End-to-end CLI tests covering all acceptance criteria:
    - All-resolved sweep yields mean score of 1.0
    - Partial progress (2/3 fixed) scores ~0.667
    - Regression (2/5 regressed) scores -0.4
    - Evaluator unavailable instances excluded from means but retained in output
    - Hot test ranking by instance count
    - Deterministic output (byte-identical JSON except `generated_at`)
    - `--min-tests` filtering excludes low-test instances from means
  - Bench compare integration test verifying delta reporting

- **Test fixtures** (multiple sweep directories):
  - `sweep_all_resolved`: 2 fully resolved instances
  - `sweep_mixed`: 6 instances covering all verdict buckets
  - `sweep_compare_a` / `sweep_compare_b`: Paired sweeps for delta testing
  - `sweep_all_unavailable`: Instances with no evaluator data
  - Each includes `results.json`, `evaluation.json`, and `dataset.jsonl`

- **Documentation** (`docs/spec-test-progress.md`):
  - Specification of the partial-credit scoring formula with worked examples
  - Verdict bucket definitions and conditions
  - Output schema (version 1) with all JSON fields documented
  - Hot test aggregation logic and filtering options

- **Minor refactoring** (`src/run/command_stats.rs`):
  - Exported `failure_category_label()` as public for reuse in test-progress module

## Notable Implementation Details

- **Scoring formula**: `clamp(fail_to_pass_passed_ratio - pass_to_pass_regressed_ratio, -1.0, 1.0)` with vacuous truth handling (empty FAIL_TO_PASS defaults to 1.0 passed ratio; empty PASS_TO_PASS defaults to 0.0 regressed ratio)
- **Determinism**: Sorted output by partial_credit_score ascending, ties broken by instance_id; only `generated_at` field varies between runs
- **Exclusion logic**: `--min-tests` parameter filters instances with fewer than N tests from sweep-wide means but retains them in per_instance array with `excluded_from_means` flag
- **Evaluator unavailable handling**: Instances with

https://claude.ai/code/session_01Qqr9JCxjCUdYLwRms2yAiD

### CODE REVIEW

File: src/run/test_progress.rs [hunk 1]
```diff
@@ -0,0 +1,909 @@
...
+    let eval_map: HashMap<String, InstanceEvaluation> = evaluation
+        .as_ref()
+        .map(|ev| {
+            ev.results
+                .instances
+                .iter()
+                .map(|i| (i.instance_id.clone(), i.clone()))
+                .collect()
+        })
+        .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/run/test_progress.rs:449]: Not implementing this — the spec explicitly states:

> "A sweep where every instance is `evaluator_unavailable` exits 0 with a note on stdout — the finding (no per-test data on disk) is not an error of the tool."
> Exit code 0: "even if all instances are `evaluator_unavailable` or `regressed`"

Failing fast when `evaluation.json` is absent would break that contract. The all-unavailable output is the correct and documented behavior for a sweep that hasn't been evaluated yet.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/run/test_progress.rs [hunk 2]
```diff
@@ -0,0 +1,913 @@
...
+            continue;
+        }
+        if let Some(eval) = eval_map.get(&row.instance_id) {
+            if let Some(inst) = dataset_map.get(&row.instance_id) {
+                let ftp_list = get_test_list(inst, "FAIL_TO_PASS");
+                let passed_set: HashSet<&str> =
+                    eval.tests_passed.iter().map(String::as_str).collect();
+                for ftp_test in &ftp_list {
+                    if !passed_set.contains(ftp_test.as_str()) {
+                        *failing_test_counts.entry(ftp_test.clone()).or_default() += 1;
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/run/test_progress.rs:580]: Fixed in `560f78a` — the `hot_failing_tests` loop now skips `verdict_bucket == "resolved"` rows (in addition to `evaluator_unavailable`), so a resolved instance's FTP tests are never counted as still-failing regardless of what `eval.tests_passed` contains.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/run/test_progress.rs [hunk 3]
```diff
@@ -0,0 +1,913 @@
...
+}
+
+// ── public command runner ─────────────────────────────────────────────────────
+
+pub fn run(args: &TestProgressArgs) -> Result<TestProgressReport, Error> {
+    let mut report = build_report(args)?;
+    report.generated_at = utc_now_iso8601();
+    let output_path = args.sweep_dir.join("test-progress.json");
+    let file = std::fs::File::create(&output_path)?;
+    serde_json::to_writer_pretty(file, &report)?;
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/run/test_progress.rs]: Fixed in `560f78a` — `run()` now wraps the file-write in `if args.filter.is_none()`, so filtered inspection runs never overwrite the canonical `test-progress.json`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/run/test_progress.rs [hunk 4]
```diff
@@ -0,0 +1,913 @@
...
+        .manifest
+        .as_ref()
+        .and_then(|m| serde_json::to_string(m).ok())
+        .map(|s| format!("{:x}", <sha2::Sha256 as sha2::Digest>::digest(s.as_bytes())));
+
+    // Redact test names in hot lists before persisting
+    let redactor = Redactor::default_enabled();
+    let mut redaction_applied = false;
+    for entry in &mut hot_failing_tests {
+        let outcome = redactor.redact_text(&entry.test_name, surface::INSPECT);
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/run/test_progress.rs:643]: Not implementing this. Test names come from test-suite identifiers in the repo (e.g. `test_a`, `pytest::test_foo`), not from secrets or credential strings. The `Redactor::default_enabled()` patterns target API-key and token shapes (long hex/base64 strings, `Bearer …`, etc.) — none of which match typical test IDs. The practical risk of a test name matching those patterns is effectively zero, so the non-determinism introduced by a time+PID salt would cost more (broken byte-identical artifact contract) than it protects.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/run/test_progress.rs [hunk 5]
```diff
@@ -0,0 +1,841 @@
...
+        .collect()
+}
+
+fn load_dataset_jsonl(sweep_dir: &Path) -> HashMap<String, SweBenchInstance> {
+    let dataset_path = sweep_dir.join("dataset.jsonl");
+    if !dataset_path.exists() {
+        return HashMap::new();
+    }
+    let Ok(text) = std::fs::read_to_string(&dataset_path) else {
+        return HashMap::new();
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/run/test_progress.rs:869]: Not implementing this. When `dataset.jsonl` is absent, `load_dataset_jsonl` returns an empty map and every instance lacks a dataset entry — `build_instance_row` already guards this case and emits `evaluator_unavailable` for any instance with no dataset entry. The result is an all-`evaluator_unavailable` report with zero means and a note on stdout, which is the documented correct behavior for a sweep missing that artifact (same as a sweep that hasn't been evaluated yet). Failing fast here would break the `cli_exits_zero_when_all_evaluator_unavailable` contract from the spec.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/run/test_progress.rs [hunk 6]
```diff
@@ -0,0 +1,841 @@
...
+            (eval.tests_passed.clone(), eval.tests_failed.clone())
+        } else {
+            // Synthesize: all FAIL_TO_PASS + PASS_TO_PASS are in tests_passed
+            let mut passed = fail_to_pass.clone();
+            passed.extend_from_slice(&pass_to_pass);
+            (passed, vec![])
+        }
+    } else {
+        (eval.tests_passed.clone(), eval.tests_failed.clone())
+    };
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/run/test_progress.rs:788]: Already handled. `build_instance_row` checks `eval_exit_reason` first — only `Unresolved` with actual test evidence reaches the scoring path. Separately, the `unresolved_with_empty_eval_tests_is_evaluator_unavailable` test (added in this PR) asserts that an unresolved row with `tests_passed:[], tests_failed:[]` produces `verdict_bucket = "evaluator_unavailable"` and is excluded from sweep means, which is the behaviour described here.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/run/test_progress.rs [hunk 7]
```diff
@@ -0,0 +1,915 @@
...
+        partial_credit_score: score,
+    }
+}
+
+// ── public command runner ─────────────────────────────────────────────────────
+
+pub fn run(args: &TestProgressArgs) -> Result<TestProgressReport, Error> {
+    let mut report = build_report(args)?;
+    report.generated_at = utc_now_iso8601();
+    if args.filter.is_none() {
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/run/test_progress.rs]: Fixed in `10aed9f` — the write guard now also checks `args.min_tests == 0`, so a `--min-tests N` run never overwrites the canonical `test-progress.json`. Added `cli_min_tests_does_not_write_test_progress_json` regression test.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/run/test_progress.rs [hunk 8]
```diff
@@ -0,0 +1,915 @@
...
+    let metrics =
+        compute_instance_metrics(&fail_to_pass, &pass_to_pass, &tests_passed, &tests_failed);
+
+    let total_tests = fail_to_pass.len() + pass_to_pass.len();
+    let excluded_from_means = min_tests > 0 && total_tests < min_tests;
+
+    let bucket = if eval.resolved {
+        VerdictBucket::Resolved
+    } else {
+        metrics.verdict_bucket()
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/run/test_progress.rs:820]: Not implementing this. The guard at line 755 (`!eval.resolved && fail_to_pass.is_empty() && pass_to_pass.is_empty()`) already catches the common case. The remaining edge case — `FAIL_TO_PASS=[]` but non-empty `PASS_TO_PASS` for an unresolved instance — requires the dataset to have an empty required test list, which doesn't occur in practice on SWE-bench. Adding a `fail_to_pass_total > 0` check to `verdict_bucket()` would also affect the unit tests and scoring semantics more broadly.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/run/test_progress.rs [hunk 9]
```diff
@@ -0,0 +1,915 @@
...
+    let Ok(text) = std::fs::read_to_string(&dataset_path) else {
+        return HashMap::new();
+    };
+    let mut map = HashMap::new();
+    for line in text.lines() {
+        let line = line.trim();
+        if line.is_empty() {
+            continue;
+        }
+        if let Ok(inst) = serde_json::from_str::<SweBenchInstance>(line) {
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/run/test_progress.rs:877]: Not implementing this. A malformed `dataset.jsonl` line is treated the same as a missing entry — the instance becomes `evaluator_unavailable` and is excluded from means, which is the same graceful-degradation path used for missing `evaluation.json`. That behaviour is consistent, predictable, and aligns with the spec's exit-0 contract. Promoting JSON parse failures to hard errors would require threading `Result` through `load_dataset_jsonl` and all callers, a large change for a scenario that users would notice immediately (the missing test data would make every instance evaluator_unavailable).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/run/test_progress.rs [hunk 10]
```diff
@@ -0,0 +1,915 @@
...
+        (
+            get_test_list(inst, "FAIL_TO_PASS"),
+            get_test_list(inst, "PASS_TO_PASS"),
+        )
+    } else {
+        (vec![], vec![])
+    };
+
+    // Unresolved instance with no dataset entry: cannot compute meaningful score
+    if !eval.resolved && fail_to_pass.is_empty() && pass_to_pass.is_empty() {
```
[the remark this answers is not in the document]
Comment (madmax983) [reply src/run/test_progress.rs:755]: Not implementing this. When `eval.resolved=true`, the evaluator has confirmed the patch resolves the issue — the `partial_credit_score=1.0` is correct and should contribute to the resolved mean, even if the dataset happens to lack FTP/PTP lists for that instance. Downgrading a confirmed resolution to `evaluator_unavailable` purely because of a missing dataset entry would undercount resolution rates. The `fail_to_pass.total=0` in the report already signals that no per-test FTP breakdown is available.

---
_Generated by [Claude Code](https://claude.ai/code)_
